src/Storages/StorageSnapshot.cpp

```diff
@@ -112,21 +112,34 @@ NameAndTypePair StorageSnapshot::getColumn(const GetColumnsOptions & options, co
     return *column;
 }
 
-Block StorageSnapshot::getSampleBlockForColumns(const Names & column_names) const
+Block StorageSnapshot::getSampleBlockForColumns(const Names & column_names,const NameToNameMap & parameter_values) const
 {
     Block res;
+
     const auto & columns = getMetadataForQuery()->getColumns();
     for (const auto & name : column_names)
     {
-        auto column = columns.tryGetColumnOrSubcolumn(GetColumnsOptions::All, name);
-        auto object_column = object_columns.tryGetColumnOrSubcolumn(GetColumnsOptions::All, name);
+        std::string column_name  = name;
+        std::string substituted_column_name = name;
+        std::string::size_type pos = 0u;
+        for (auto parameter : parameter_values)
+        {
+            if ((pos = substituted_column_name.find("_CAST(" + parameter.second)) != std::string::npos)
```

Comment:
This place confuses me a bit. So it is for the case `CREATE VIEW pv6 AS SELECT Price+{price:UInt64} ...`? and we need to make a substitution because otherwise `tryGetColumnOrSubcolumn` method will fail? But if we added a cast earlier, it seems like cast should already be in `columns`'s column name and `tryGetColumnOrSubcolumn` will succeed. So may be we just need to move adding cast earlier so we do not need to replace here?

Reply:
yes, you are right, it is for such queries and we need to make substitution otherwise `tryGetColumnOrSubcolumn` will fail.
Sorry I didnt get at what stage you suggest we add the cast so that we can do not replace here?
The concern is we have plus(Price, price) in source columns, (CAST in expected column) and we want to keep this in source columns because this create+select query can be used again with different substitutes in future queries.
let me know if you have any better suggestions to do this.

Reply:
thanks, now it became clear